File: medtrue_app/lib/models/product_models.dart
class Product {
  final int? productId; // Nullable for new products
  final String name;
  final String? packingDesc;
  final String? barcode;
  final int? companyId;
  final int? saltId;
  final int? categoryId;
  final int? unitPrimaryId;
  final int? unitSecondaryId;
  final double conversionFactor;
  final double mrp;
  final double purchaseRate;
  final double costRate;
  
  // Images
  final List<ProductImage>? images;

  Product({
    this.productId,
    required this.name,
    this.packingDesc,
    this.barcode,
    this.companyId,
    this.saltId,
    this.categoryId,
    this.unitPrimaryId,
    this.unitSecondaryId,
    this.conversionFactor = 1.0,
    this.mrp = 0.0,
    this.purchaseRate = 0.0,
    this.costRate = 0.0,
    this.images,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      productId: json['productId'],
      name: json['name'],
      packingDesc: json['packingDesc'],
      barcode: json['barcode'],
      companyId: json['companyId'],
      saltId: json['saltId'],
      categoryId: json['categoryId'],
      unitPrimaryId: json['unitPrimaryId'],
      unitSecondaryId: json['unitSecondaryId'],
      conversionFactor: (json['conversionFactor'] as num?)?.toDouble() ?? 1.0,
      mrp: (json['mrp'] as num?)?.toDouble() ?? 0.0,
      purchaseRate: (json['purchaseRate'] as num?)?.toDouble() ?? 0.0,
      costRate: (json['costRate'] as num?)?.toDouble() ?? 0.0,
      // Handle Image parsing later manually or via extra field
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'productId': productId,
      'name': name,
      'packingDesc': packingDesc,
      'barcode': barcode,
      'companyId': companyId,
      'saltId': saltId,
      'categoryId': categoryId,
      'unitPrimaryId': unitPrimaryId,
      'unitSecondaryId': unitSecondaryId,
      'conversionFactor': conversionFactor,
      'mrp': mrp,
      'purchaseRate': purchaseRate,
      'costRate': costRate,
    };
  }
}

class ProductImage {
  final int? imgId;
  final int? productId;
  final String imagePath;
  final bool isPrimary;

  ProductImage({
    this.imgId,
    this.productId,
    required this.imagePath,
    this.isPrimary = false,
  });

  factory ProductImage.fromJson(Map<String, dynamic> json) {
    return ProductImage(
      imgId: json['imgId'],
      productId: json['productId'],
      imagePath: json['imagePath'],
      isPrimary: json['isPrimary'] ?? false,
    );
  }
}

File: Models/ProductModels.cs
namespace MedTrueApi.Models;

public class Product
{
    public long ProductId { get; set; }
    public string Name { get; set; } = string.Empty;
    public string? PackingDesc { get; set; }
    public string? Barcode { get; set; }

    // Foreign Keys
    public int? CompanyId { get; set; }
    public int? SaltId { get; set; }
    public int? CategoryId { get; set; }
    public string? HsnCode { get; set; }
    public int? ItemTypeId { get; set; }

    // Units
    public int? UnitPrimaryId { get; set; }
    public int? UnitSecondaryId { get; set; }
    public decimal ConversionFactor { get; set; } = 1;
    public string? UnitPrimaryName { get; set; } // Fetched via JOIN

    // Status
    public string Status { get; set; } = "CONTINUE";
    public bool IsHidden { get; set; }
    public bool IsDecimalAllowed { get; set; }
    public bool HasPhoto { get; set; }

    // Regulatory
    public bool IsNarcotic { get; set; }
    public int? ScheduleHId { get; set; }

    // Inventory
    public string? RackNumber { get; set; }
    public int MinQty { get; set; }
    public int MaxQty { get; set; }
    public int ReorderQty { get; set; }
    public bool AllowNegativeStock { get; set; }
    public int CurrentStock { get; set; } // Added

    // Pricing
    public decimal Mrp { get; set; }
    public decimal PurchaseRate { get; set; }
    public decimal CostRate { get; set; }
    public decimal SalePrice { get; set; } // Added

    // Tax
    public decimal SgstPercent { get; set; }
    public decimal CgstPercent { get; set; }
    public decimal IgstPercent { get; set; }

    // Discount
    public decimal ItemDiscount1 { get; set; }
    public decimal SpecialDiscount { get; set; }
    public decimal MaxDiscountPercent { get; set; }
    public decimal SaleMargin { get; set; }

    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }
}

public class ProductBatch
{
    public long BatchId { get; set; }
    public long ProductId { get; set; }
    public string BatchNumber { get; set; } = string.Empty;
    public DateOnly ExpiryDate { get; set; }
    public decimal? Mrp { get; set; }
    public decimal? PurchaseRate { get; set; }
    public decimal? QuantityAvailable { get; set; }
    public DateOnly EntryDate { get; set; }
}

public class ProductImage
{
    public int ImgId { get; set; }
    public long ProductId { get; set; }
    public string ImagePath { get; set; } = string.Empty;
    public bool IsPrimary { get; set; }
    public int DisplayOrder { get; set; }
    public DateTime CreatedAt { get; set; }
}
